### ISSUE TITLE
Visual Studio 15.6 deadlocks on solution load for .NET Core SDK project when a <None Include=""> is added to file list

### ISSUE THREAD

**Issue by btrepp** [Microsoft/visualfsharp#4300]
(Edited by @cartermp to update repro steps)

15.6 crashes on load of this solution 
https://github.com/btrepp/vspreviewcrash

#### Repro steps

1. git clone `https://github.com/btrepp/vspreviewcrash`
2. `cd vspreviewcrash\dotnet`
3. `.paket\paket.exe install`
4. `dotnet restore`
5. Open solution file.

#### Expected behavior

Solution loads.

#### Actual behavior

A popup with "a non fatal error has occured" appears, but then VS locks up

#### Known workarounds

None

#### Related information 

Operating system is windows 10, Machine had brand new windows installed and then Visual studio installed.

**Comment by cartermp** [comment]
Sorry not at computer - can you remove the FSharp.NET.Sdk attribute and
see what happens?

On Thu, Feb 1, 2018, at 17:38, btrepp wrote:
> 15.6 crashes on load of this solution
> https://github.com/btrepp/vspreviewcrash


> Repro steps


>  1. Clone repo
>  2. Open solution in 15.6 preview 3
> Expected behavior


> Solution loads like in 15.5.6


> Actual behavior


> A popup with "a non fatal error has occured" appears, but then VS
> locks up> Known workarounds


> None


> Related information


> Operating system is windows 10, Machine had brand new windows
> installed and then Visual studio installed.> — You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub[1], or mute the
> thread[2].> 


Links:

  1. https://github.com/Microsoft/visualfsharp/issues/4300

**Comment by btrepp** [comment]
That helps with the solution level crash, the IDE loads. But fails to load the project with
"The project file is incomplete. Expected imports are missing"

Its nice enough to create a log file for this. Which contains

```
=====================
2/02/2018 10:52:10 AM
Recoverable
System.AggregateException: One or more errors occurred. ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.VisualStudio.ProjectSystem.UpToDate.BuildUpToDateCheck.Initialize()
   at Microsoft.VisualStudio.ProjectSystem.OnceInitializedOnceDisposed.EnsureInitialized(Boolean allowInitializing)
   at Microsoft.VisualStudio.ProjectSystem.UpToDate.BuildUpToDateCheck.Load()
   at Microsoft.VisualStudio.ProjectSystem.ConfiguredProjectImpl.<>c__DisplayClass108_2.<<InitializeAsync>b__6>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.ProjectMultiThreadedService.<>c__DisplayClass28_1.<<Fork>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.VisualStudio.ProjectSystem.UpToDate.BuildUpToDateCheck.Initialize()
   at Microsoft.VisualStudio.ProjectSystem.OnceInitializedOnceDisposed.EnsureInitialized(Boolean allowInitializing)
   at Microsoft.VisualStudio.ProjectSystem.UpToDate.BuildUpToDateCheck.Load()
   at Microsoft.VisualStudio.ProjectSystem.ConfiguredProjectImpl.<>c__DisplayClass108_2.<<InitializeAsync>b__6>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.ProjectMultiThreadedService.<>c__DisplayClass28_1.<<Fork>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)<---

===================
```

**Comment by KevinRansom** [comment]
@Pilchie @jmarolf 

That stack looks like Project System or CPS.

**Comment by cartermp** [comment]
Tagging as external for now, will update or close as-appropriate

**Comment by Pilchie** [comment]
Tagging @panopticoncentral - do you know how this could happen?

**Comment by panopticoncentral** [comment]
All we do in Initialize is hook up to the dataflow. Best guess is that some condition is causing one of the CPS services we expect to be there to be null. We can harden against this, but it's surprising that a service wouldn't be there. Maybe the import service since imports are mentioned above? I'll take a look in the debugger.

**Comment by cartermp** [comment]
Big title/text update for tracking purposes.

Basically, if a `<None Include-"">` is in the set of files for compilation in a .NET Core SDK project, VS will deadlock. This is because default file globbing for the project system will pick it up anyways, and this ultimately leads to duplicate items referring to the same file.

The workaround is to apply this in the project file:

```xml
  <ItemGroup>
    <None Remove="*.*" />
  </ItemGroup>
```

Then the F# project files with `None Include` items will work just fine. A fix is underway in the project system.

**Comment by btrepp** [comment]
Depending on your projects (eg I used folders). The nuclear workaround is
```
  <ItemGroup>
    <None Remove="**/*" />
  </ItemGroup>
```
Which throws all the automagic stuff in the bin and makes it so you explicitly include items. Adding this everywhere brought my solution back to being use-able.

**Comment by cartermp** [comment]
The fix for this has been merged in and will be available in a servicing release for 15.6 soon.
